
### Pandas Data Cleaning

Load data.


In [1]:
# Importing Libraries
import pandas as pd
from datasets import load_dataset
import matplotlib.pyplot as plt  

# Loading Data
dataset = load_dataset('lukebarousse/data_jobs')
df = dataset['train'].to_pandas()

# Data Cleanup
df['job_posted_date'] = pd.to_datetime(df['job_posted_date'])


### Handling Missing Data
#### Review

This is what we learned in the basics section, this is just a refresher of how we've handled null values before. Feel free to skip this.

#### Notes

   - df.dropna(): Drop missing values.

#### Examples

Here we are only drop values if all of their values are missing.


In [2]:
df_cleaned = df.dropna(how='all')
df_cleaned.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 785741 entries, 0 to 785740
Data columns (total 17 columns):
 #   Column                 Non-Null Count   Dtype         
---  ------                 --------------   -----         
 0   job_title_short        785741 non-null  object        
 1   job_title              785740 non-null  object        
 2   job_location           784696 non-null  object        
 3   job_via                785733 non-null  object        
 4   job_schedule_type      773074 non-null  object        
 5   job_work_from_home     785741 non-null  bool          
 6   search_location        785741 non-null  object        
 7   job_posted_date        785741 non-null  datetime64[ns]
 8   job_no_degree_mention  785741 non-null  bool          
 9   job_health_insurance   785741 non-null  bool          
 10  job_country            785692 non-null  object        
 11  salary_rate            33067 non-null   object        
 12  salary_year_avg        22003 non-null   floa



Right now all we can do is drop values if they're missing. But that's not useful right now because our DataFrame didn't have any.

So, what if we wanted to fill the missing values with something else? This is expecially useful so we don't run into errors when dealing with NaN values.

#### Fillna

#### Notes

   - df.fillna(): Fill missing values

#### Examples

Let's fill in instances where there's no salary info (aka these columns have NaN values salary_rate, salary_year_avg, salary_hour_avg) with 0.

We're going to look at a few rows in these 3 columns right now, so we can compare what we've done before to after.

We'll use iloc to look at the first 10 rows :20 and the salary information rows 11:14.


In [3]:
df_cleaned.iloc[:10,11:14]

salary_rate  salary_year_avg  salary_hour_avg
0        None              NaN              NaN
1        None              NaN              NaN
2        None              NaN              NaN
3        None              NaN              NaN
4        None              NaN              NaN
5        None              NaN              NaN
6        None              NaN              NaN
7        None              NaN              NaN
8        None              NaN              NaN
9        None              NaN              NaN

We fill the values for the 3 columns with 0 using fillna().

In [4]:
fill_values = ['salary_rate', 'salary_year_avg', 'salary_hour_avg']
df_filled = df_cleaned.fillna(0)
df_filled.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 785741 entries, 0 to 785740
Data columns (total 17 columns):
 #   Column                 Non-Null Count   Dtype         
---  ------                 --------------   -----         
 0   job_title_short        785741 non-null  object        
 1   job_title              785741 non-null  object        
 2   job_location           785741 non-null  object        
 3   job_via                785741 non-null  object        
 4   job_schedule_type      785741 non-null  object        
 5   job_work_from_home     785741 non-null  bool          
 6   search_location        785741 non-null  object        
 7   job_posted_date        785741 non-null  datetime64[ns]
 8   job_no_degree_mention  785741 non-null  bool          
 9   job_health_insurance   785741 non-null  bool          
 10  job_country            785741 non-null  object        
 11  salary_rate            785741 non-null  object        
 12  salary_year_avg        785741 non-null  floa

Now if we compare the results using iloc again on our new DataFrame. We see that the previous NaN values in the columns ( salary_rate, salary_year_avg, salary_hour_avg) have been replaced with 0.

In [5]:
df_filled.iloc[:10,11:14]

salary_rate  salary_year_avg  salary_hour_avg
0           0              0.0              0.0
1           0              0.0              0.0
2           0              0.0              0.0
3           0              0.0              0.0
4           0              0.0              0.0
5           0              0.0              0.0
6           0              0.0              0.0
7           0              0.0              0.0
8           0              0.0              0.0
9           0              0.0              0.0


### Drop Duplicates

#### Notes

   - drop_duplicates(): Remove duplicate rows.
   - Analysts will often need to clean up data and one of the most common issues we run into is duplicate values.

#### Examples

Now that we've dealt with NaN values. Let's continue cleaning the data by removing any duplicate rows.


In [6]:
df_unique = df_filled.drop_duplicates()
df_unique.info()

<class 'pandas.core.frame.DataFrame'>
Index: 785640 entries, 0 to 785740
Data columns (total 17 columns):
 #   Column                 Non-Null Count   Dtype         
---  ------                 --------------   -----         
 0   job_title_short        785640 non-null  object        
 1   job_title              785640 non-null  object        
 2   job_location           785640 non-null  object        
 3   job_via                785640 non-null  object        
 4   job_schedule_type      785640 non-null  object        
 5   job_work_from_home     785640 non-null  bool          
 6   search_location        785640 non-null  object        
 7   job_posted_date        785640 non-null  datetime64[ns]
 8   job_no_degree_mention  785640 non-null  bool          
 9   job_health_insurance   785640 non-null  bool          
 10  job_country            785640 non-null  object        
 11  salary_rate            785640 non-null  object        
 12  salary_year_avg        785640 non-null  float64  



If you compare that with the original DataFrame which had 785741 entries. This new DataFrame df_unique has 785640 entries. It removed 101 entries.

Now let's see what would happen if we tried to remove duplicates from job_title.


In [7]:
df_unique = df_filled.drop_duplicates(subset=['job_title'])
df_unique.info()

<class 'pandas.core.frame.DataFrame'>
Index: 234674 entries, 0 to 785740
Data columns (total 17 columns):
 #   Column                 Non-Null Count   Dtype         
---  ------                 --------------   -----         
 0   job_title_short        234674 non-null  object        
 1   job_title              234674 non-null  object        
 2   job_location           234674 non-null  object        
 3   job_via                234674 non-null  object        
 4   job_schedule_type      234674 non-null  object        
 5   job_work_from_home     234674 non-null  bool          
 6   search_location        234674 non-null  object        
 7   job_posted_date        234674 non-null  datetime64[ns]
 8   job_no_degree_mention  234674 non-null  bool          
 9   job_health_insurance   234674 non-null  bool          
 10  job_country            234674 non-null  object        
 11  salary_rate            234674 non-null  object        
 12  salary_year_avg        234674 non-null  float64  

If we look at it now. It looks like we removed quite a few rows. Now all of these rows have unique job_titles.

In [8]:
df_unique.head(10)

job_title_short                                          job_title  \
0  Senior Data Engineer  Senior Clinical Data Engineer / Principal Clin...   
1          Data Analyst                                       Data Analyst   
2         Data Engineer  Data Engineer/Scientist/Analyst, Mid or Senior...   
3         Data Engineer  LEAD ENGINEER - PRINCIPAL ANALYST - PRINCIPAL ...   
4         Data Engineer                             Data Engineer- Sr Jobs   
5         Data Engineer                                  GCP Data Engineer   
6  Senior Data Engineer                  Senior Data Engineer  - GCP Cloud   
7         Data Engineer                                      Data Engineer   
8      Business Analyst           Technology & Operations Business Analyst   
9        Data Scientist                                  Data Scientist II   

                   job_location             job_via         job_schedule_type  \
0                 Watertown, CT     via Work Nearby                 Full-time   
1  Guadalajara, Jalisco, Mexico    via BeBee México                 Full-time   
2               Berlin, Germany        via LinkedIn                 Full-time   
3               San Antonio, TX   via Diversity.com                 Full-time   
4                Washington, DC  via Clearance Jobs                 Full-time   
5                      Anywhere    via ZipRecruiter  Contractor and Temp work   
6                  Dearborn, MI        via LinkedIn                 Full-time   
7                      Anywhere        via LinkedIn                 Full-time   
8           Copenhagen, Denmark     via Trabajo.org                 Full-time   
9                      Anywhere    via ZipRecruiter                 Full-time   

   job_work_from_home          search_location     job_posted_date  \
0               False     Texas, United States 2023-06-16 13:44:15   
1               False                   Mexico 2023-01-14 13:18:07   
2               False                  Germany 2023-10-10 13:14:55   
3               False     Texas, United States 2023-07-04 13:01:41   
4               False                    Sudan 2023-08-07 14:29:36   
5                True                  Georgia 2023-11-07 14:01:59   
6               False   Florida, United States 2023-03-27 13:18:18   
7                True                  Romania 2023-12-07 13:40:49   
8               False                  Denmark 2023-06-05 13:44:34   
9                True  New York, United States 2023-04-23 13:02:57   

   job_no_degree_mention  job_health_insurance    job_country salary_rate  \
0                  False                 False  United States           0   
1                  False                 False         Mexico           0   
2                  False                 False        Germany           0   
3                   True                 False  United States           0   
4                  False                 False          Sudan           0   
5                  False                 False  United States           0   
6                  False                 False  United States           0   
7                  False                 False        Romania           0   
8                  False                 False        Denmark           0   
9                  False                 False  United States           0   

   salary_year_avg  salary_hour_avg                   company_name  \
0              0.0              0.0           Boehringer Ingelheim   
1              0.0              0.0     Hewlett Packard Enterprise   
2              0.0              0.0       ALPHA Augmented Services   
3              0.0              0.0   Southwest Research Institute   
4              0.0              0.0                Kristina Daniel   
5              0.0              0.0                smart folks inc   
6              0.0              0.0  Miracle Software Systems, Inc   
7              0.0              0.0                          Zitec   
8       

For our example we don't really need to remove any duplicates right now, but it's important to understand the concept.